proposals/0000-th-stage-hygiene.rst

```diff
@@ -0,0 +1,397 @@
+  Stage 2 - 1 = Stage 1.]
+
+  Switching all existing ``derive-*`` code generators to TH would probably make them lighter and easier to maintain.
+  It should also allow building GHC the binary with plain `cabal`.
+  Hadrian would be one step closer to being another implementation of Cabal/cabal-install without GHC-specific logic.
+
+``*_BUILD_*`` and ``*_HOST_*`` not always defined
+  The conditional definition of the CPP macros ensures they don't pollute the purity of the build when they don't matter.
+  This is important for highly pure build systems like Nix to not have to needless rebuild stuff when the target platform changes.
+  It will also cut down on people improperly using "target" when they meant "host".
+
+``Lift`` and qualified goals
+  When we carefully introduce lifting to stage-hygienic goals, we need to ensure that the type being lifted is the same or "close enough".
+  This means we will need to introduce a intra-package constraint on package defining that type across the stage pair where Lift is made available.
+  Conceptually, there might be an auto-generated package with the orphan ``Lift`` instance which imposes the same version constraint on it's library dependency in both stages.
+
+  In particular, existing qualified dependencies from ``setup-depends`` and ``build-tool-depends`` are from stage *n* to *n - 1*;
+  that the stages are different alone explains why versions are allowed to differ.
+  In particular this means given a dependency edge where the needed and needing components are in the same package regardless of their relative stage indices,
+  the same version of the package must be used for both.
+
+
+Costs and Drawbacks
+-------------------
+
+- This is a huge amount of work.
+  But I am fine chipping away it over a long period of time.
+
+- Even a temporary conflict between typed TH and this could slow typed TH's adoption.
+
+- I don't know of precedent for extensions that prevent modules from being linked together.
+
+- Most existing libraries with commonly used TH helpers (`lens`, `aeson`) have the TH in the same call component but in a different module.
+  To leverage this proposal, we would have to refactor them to put those modules in a separate library component.
+  It would take decent amount of conditional code to still support old GHCs, and even more to not be a breaking change on those old libraries.
+
+Alternatives
+------------
+
+There is no fundamental reason modules couldn't export non-stage-0 items, and libraries expose non-stage-0 modules.
+At the cost of more complexity, there could be a `.lib` or `.so` for each exposed stage, and imports would be offset to match the ``#import <offset>`` literal.
+But in fairness, this might allow a smoother transition form how libraries are structured today.
+For example, one could do ``#import 1 Control.Lens.Lens`` in ``Control.Lens.TH`` while exposing ``Control.Lens.TH`` from the same library just like today.
+I decided against this as a matter of taste.
```

Comment:
I think this also makes cabal's job easier.

Reply:
Good point.